### ISSUE TITLE
Prototype a generic Skill Memory mechanism for continual learning

### ISSUE THREAD

**Issue by kobros-tech** [kobros-tech/avalanche#1]
## Goal

Investigate whether the reusable mechanism developed in [`kobros-tech/skill-memory`](https://github.com/kobros-tech/skill-memory) can be integrated cleanly into Avalanche as a **generic continual-learning capability**, rather than as an arithmetic-specific feature.

This issue is an implementation/research plan. The outcome should be a working prototype in this fork, backed by tests and benchmark evidence. If the result is useful, clean, and consistent with Avalanche's architecture, prepare a focused PR for upstream [`ContinualAI/avalanche`](https://github.com/ContinualAI/avalanche).

## Background

`skill-memory` separates the reusable mechanism from research experiments. The mechanism includes skill representations, a registry, compatibility scoring, independent identification/solve logic, and dynamic acquisition/reuse.

The important idea for Avalanche is **not arithmetic**. The candidate contribution is a generic mechanism where a learner can:

1. acquire a reusable skill;
2. store it in skill memory;
3. identify whether a new experience can reuse an existing skill;
4. reuse compatible knowledge when appropriate;
5. otherwise learn the new skill normally;
6. register the resulting skill for future experiences.

## Target architecture

Do not copy `skill-memory` wholesale into Avalanche. First map its concepts onto Avalanche's existing abstractions (`Strategy`, `Experience`, models, plugins, training loop, evaluation, etc.).

```text
new experience
      |
      v
identify / characterize task
      |
      v
query skill memory
      |
      +-----------------------------+
      |                             |
compatible reusable skill       no useful skill
      |                             |
      v                             v
reuse / initialize             normal acquisition
      |                             |
      +--------------+--------------+
                     |
                     v
               train / evaluate
                     |
                     v
              register new skill
```

The implementation should be model-agnostic as far as Avalanche's abstractions permit. Arithmetic is only a proof-of-concept/test case, not the definition of the public API.

### Phase 0 — Baseline

- Work from the current `master` branch.
- Record the exact upstream Avalanche revision being targeted.
- Run relevant existing tests before modifications.
- Inspect current strategies, plugins, model abstractions, evaluation utilities, examples, and training lifecycle.
- Identify the smallest appropriate integration point; do not build a parallel training framework.

### Phase 1 — Architecture mapping

Document the mapping between `skill-memory` and Avalanche:

| skill-memory concept | Avalanche candidate abstraction |
|---|---|
| Skill | reusable learned component/state associated with an experience/task |
| SkillRegistry | persistent registry owned by a strategy or dedicated memory object |
| compatibility() | task/skill compatibility estimator |
| identify() | task/skill identification mechanism |
| DynamicSkillController | Avalanche strategy/plugin orchestration |
| independent solve gate | evaluation/probing mechanism without test-set leakage |

Validate these mappings against the actual Avalanche APIs before implementation.

### Phase 2 — Minimal generic API

Design the smallest public interface supporting the mechanism, conceptually:

```python
memory = SkillMemory(...)
memory.register(skill)
skill = memory.retrieve(task)
score = memory.compatibility(skill, task)
```

Only add operations justified by concrete use cases/tests.

Requirements:

- no arithmetic-specific names in the generic core;
- deterministic behavior where practical;
- explicit compatibility/reuse decision;
- clear no-match behavior;
- no hidden access to future experiences or test data;
- clean serialization/checkpoint behavior if required.

### Phase 3 — Avalanche-native strategy/plugin

Implement the smallest working integration with Avalanche's training lifecycle. It should:

- receive a normal Avalanche experience;
- inspect/identify the experience;
- query skill memory;
- reuse an eligible skill or fall back to ordinary learning;
- train/evaluate through normal Avalanche mechanisms;
- store newly acquired skills;
- preserve registered skills unless an explicit policy permits modification.

Reuse Avalanche's optimizer, evaluator, logging, and training-loop infrastructure rather than duplicating it.

### Phase 4 — Tests

Add focused tests for:

- empty memory;
- one/multiple registered skills;
- correct compatible retrieval;
- incompatible-skill rejection;
- deterministic retrieval;
- reuse path;
- new-acquisition path;
- registration after acquisition;
- multiple sequential experiences;
- no future-information leakage;
- model/state compatibility;
- checkpoint/save/load if supported;
- normal Avalanche evaluation interaction.

Keep tests small and CI-friendly.

### Phase 5 — Arithmetic proof of concept

Port the arithmetic example from `skill-memory` only as validation:

```text
Task A -> acquire skill A
Task B -> acquire skill B
Task C -> reuse compatible prior skill when appropriate
Task D -> reject inappropriate skill and learn normally
```

Retain independent solve/evaluation logic; target test-set labels must not determine reuse.

### Phase 6 — Standard continual-learning benchmark

Before an upstream PR, evaluate the generic mechanism on at least one standard Avalanche scenario. Arithmetic alone is insufficient.

Compare appropriate baselines such as:

- ordinary training/current-state learning;
- a relevant existing Avalanche baseline;
- replay where applicable;
- Skill Memory.

Measure at minimum:

- final accuracy/performance;
- per-experience performance;
- forgetting/backward transfer where applicable;
- acquisition/training cost;
- number of reused skills;
- number of newly acquired skills;
- failed/incorrect reuse cases;
- memory and runtime overhead.

Use fixed seeds and clearly separated training/calibration/evaluation data where a reuse gate is involved.

### Phase 7 — Code/API review

Before upstream consideration:

- minimize the implementation;
- remove research-only code;
- follow Avalanche conventions;
- add appropriate typing/docstrings;
- document public APIs;
- avoid unnecessary dependencies;
- verify packaging/import behavior;
- run relevant lint/type/test checks;
- ensure existing Avalanche tests remain green.

### Phase 8 — Documentation/example

Provide one concise generic example explaining:

- what Skill Memory is;
- when reuse happens;
- what compatibility means;
- what happens with no reusable skill;
- differences from replay/regularization;
- limitations and overhead.

Arithmetic remains a demonstration; public documentation stays generic.

### Phase 9 — Upstream readiness / go-no-go

Only prepare an upstream PR if:

- the mechanism works outside arithmetic;
- the API fits Avalanche naturally;
- tests are robust;
- benchmark evidence demonstrates meaningful benefit or a clearly useful capability;
- runtime/memory overhead is documented;
- it does not duplicate an existing Avalanche feature;
- implementation is small and maintainable;
- documentation is sufficient for external users;
- relevant existing tests pass.

If not, keep it as a research/prototype contribution and document what remains unresolved rather than forcing an upstream PR.

## Design constraints

1. **Generic first:** arithmetic must not define the architecture.
2. **Avalanche-native:** reuse existing abstractions instead of creating a second framework.
3. **Small implementation:** prefer a compact mechanism and clear API.
4. **Research/mechanism separation:** benchmark scripts, plots, statistics, and paper code stay outside the reusable core.
5. **No evaluation leakage:** reuse decisions cannot depend on target test-set labels.
6. **Explicit fallback:** uncertain/incompatible skills must not be forced onto a task.
7. **Measurable benefit:** quantify whether reuse improves acquisition, performance, memory, or training cost.
8. **Reproducibility:** fixed seeds and explicit evaluation protocols.
9. **Upstream maintainability:** the final code must make sense to Avalanche contributors without knowledge of `skill-memory` or `skill-cloning`.

## Relationship to related repositories

- `skill-memory` remains the standalone reusable prototype/mechanism during development.
- `skill-cloning` remains the research project studying when learned skills transfer, when cloning accelerates acquisition, and how compatibility relates to transfer.
- This Avalanche work extracts only the generalizable continual-learning mechanism plus tests/examples.

## Definition of done

- [ ] Baseline Avalanche tests pass before implementation.
- [ ] Avalanche architecture mapped to the proposed mechanism.
- [ ] Minimal generic Skill Memory API designed.
- [ ] Avalanche-native prototype implemented.
- [ ] Unit tests added.
- [ ] Integration tests added.
- [ ] Arithmetic proof of concept works.
- [ ] At least one standard continual-learning benchmark evaluated.
- [ ] Appropriate baselines included.
- [ ] Reuse/acquisition decisions independently evaluated.
- [ ] Runtime/memory overhead measured.
- [ ] Public API/documentation cleaned up.
- [ ] Relevant existing Avalanche tests pass.
- [ ] Final upstream go/no-go decision made.

## Intended outcome

This issue is **not a commitment that the feature belongs upstream**. It is a controlled way to determine that objectively.

If the prototype demonstrates that Skill Memory is a useful, general, lightweight continual-learning capability and integrates naturally with Avalanche, the next step is a focused PR against `ContinualAI/avalanche`. Otherwise, the work remains a useful standalone/research contribution in `skill-memory` and `skill-cloning`.

## References

- `skill-memory`: https://github.com/kobros-tech/skill-memory
- `skill-cloning`: https://github.com/kobros-tech/skill-cloning
- this Avalanche fork: https://github.com/kobros-tech/avalanche
- upstream Avalanche: https://github.com/ContinualAI/avalanche

**Comment by kobros-tech** [comment]
## Initial implementation started

Created branch `feature/skill-memory-prototype` from `master` and implemented the first generic integration slice:

- `avalanche/training/skill_memory/memory.py`
  - bounded `SkillMemory`
  - immutable-by-default skill registration
  - CPU state storage to avoid retaining GPU memory
  - metadata/provenance support
  - deterministic best-match selection
  - configurable compatibility threshold
  - model state restoration
- `avalanche/training/skill_memory/plugin.py`
  - `SkillMemoryPlugin` using Avalanche's existing plugin lifecycle
  - optional compatible-skill reuse before an experience
  - normal training remains the fallback
  - post-training skill registration
  - optional optimizer-state reset on reuse
- `tests/training/test_skill_memory.py`
  - registration/copy semantics
  - capacity
  - matching/threshold behavior
  - reuse path
  - fallback path
- `avalanche/training/__init__.py`
  - exports the new primitives

This deliberately starts with the smallest Avalanche-native mechanism rather than introducing a new strategy/training loop. The next step is to validate this slice against the actual test suite, then build the arithmetic proof-of-concept and benchmark integration.

I have not marked any implementation phase as complete yet because the repository test suite has not been executed in this environment.

**Comment by kobros-tech** [comment]
Progress update: continued the prototype on `feature/skill-memory-prototype`. Added an Avalanche-native `SkillMemoryPlugin`, expanded the registry/plugin tests (threshold validation, metadata, idempotent registration, explicit replacement), and added a minimal usage example under `examples/skill_memory.py`. The implementation remains a draft; benchmark validation and real Avalanche end-to-end integration are still required before any upstream proposal.

**Comment by kobros-tech** [comment]
## Prototype status update — PR #2 / arithmetic benchmark

The prototype has now progressed beyond the original implementation notes:

### Completed / demonstrated
- [x] Generic `SkillMemory` registry implemented.
- [x] Avalanche-native `SkillMemoryPlugin` integrated through `SupervisedPlugin` lifecycle.
- [x] Focused unit tests for registration, capacity, matching, thresholds, state round-trip, reuse, fallback, and replacement policy.
- [x] Real Avalanche integration test using `Naive` and `get_fast_benchmark()`; this verifies that the plugin is invoked by the actual training lifecycle, not only through direct method calls.
- [x] Arithmetic proof-of-concept and controlled baseline-vs-Skill-Memory benchmark added.
- [x] Benchmark uses separate deterministic training/evaluation seeds and does not pass evaluation data to the compatibility function.
- [x] The benchmark demonstrates the intended decision flow: multiply→acquire, add→acquire, square→reuse:multiply, divide→acquire.

### Current benchmark interpretation
The supplied seed-7 result demonstrates that the mechanism and reuse path work, but it is **not yet evidence of overall performance benefit**. Skill Memory improves final MSE for addition and division and substantially reduces square forgetting, but final MSE is worse for multiplication and square. Therefore the result should be treated as a successful integration/proof-of-concept signal, not as an upstream-quality benchmark result.

### Remaining before upstream consideration
1. Run the arithmetic benchmark across multiple fixed seeds and report mean/variance or confidence intervals.
2. Add a standard Avalanche continual-learning scenario; arithmetic alone does not satisfy the issue's generalization requirement.
3. Compare against appropriate existing baselines (at minimum ordinary sequential training; replay or another relevant Avalanche baseline where applicable).
4. Measure training/runtime cost and memory overhead, plus reuse/acquisition counts and incorrect/failed reuse cases.
5. Review checkpoint/serialization behavior in the actual Avalanche strategy lifecycle, not only the standalone `SkillMemory.state_dict()` round trip.
6. Review API/typing/documentation and remove development-only workflow/example material that should not be part of an upstream contribution.
7. Re-run the relevant Avalanche test suite and finalize the upstream go/no-go decision.

### Important conclusion
Do **not** close this issue or prepare an upstream PR yet. PR #2 is still correctly a draft. The next highest-value step is a multi-seed benchmark followed by one standard Avalanche benchmark; only after those results should we decide whether the mechanism provides a sufficiently general and measurable benefit for upstream Avalanche.

**Comment by kobros-tech** [comment]
### Step 1.5 — arithmetic benchmark analysis cleanup

Implemented the next analysis pass on `feature/skill-memory-prototype`.

Changes:
- clarified the division experience as having no prerequisites, so its acquire decision is intentional rather than relying on a mismatched `division`/`divide` label;
- documented the exact forgetting metric: `max(0, final MSE - minimum MSE observed after any experience)`;
- added paired per-seed comparisons (`Skill Memory - baseline`) for both final MSE and forgetting;
- added mean, sample standard deviation, approximate 95% CI, and per-task win/tie counts for the paired differences;
- kept the existing 15 fixed seeds (`0..14`) and independent train/evaluation datasets;
- kept the benchmark explicitly scoped as a controlled arithmetic robustness experiment, not a general continual-learning claim.

Commits:
- `9a51b013` — clarify arithmetic prerequisites and add paired analysis
- `b56af3f` — document multi-seed paired benchmark analysis

The workflow should now be rerun so the updated `skill_memory_benchmark_multiseed.json` is produced. The next decision should be based on the paired forgetting results, followed by the standard Avalanche benchmark.

**Comment by kobros-tech** [comment]
## Step 2 started: standard Avalanche benchmark

Step 1 and Step 1.5 are complete. This PR now adds the first standard Avalanche generalization experiment:

- **Scenario:** standard `RotatedMNIST` benchmark.
- **Sequence:** rotations `[0, 30, 60, 0, 30]`, so repeated transformations create legitimate reuse opportunities without hard-coding source/target task pairs.
- **Conditions:** `Naive`, `Naive + ReplayPlugin`, and `Naive + SkillMemoryPlugin`.
- **Compatibility:** based only on the known training-task rotation metadata; evaluation samples/targets are not used for reuse decisions.
- **Metrics:** per-experience accuracy, final accuracy, forgetting on prior experiences, acquisition/reuse decisions, stored skills, and training time.
- **Initial seeds:** 0, 1, 2. This is an initial Step 2 validation; if stable and informative, the benchmark should be expanded before any broad research claim.

Implementation: `examples/skill_memory_standard_benchmark.py`

Workflow now uploads `skill-memory-rotated-mnist` containing `results/skill_memory_rotated_mnist.json`.

The arithmetic benchmark remains unchanged as the controlled mechanistic experiment.
